### PR TITLE
Add std::panic::panic_any.

### PR DESCRIPTION
**Pull request by m-ou-se**
The discussion of #67984 lead to the conclusion that there should be a macro or function separate from `std::panic!()` for throwing arbitrary payloads, to make it possible to deprecate or disallow (in edition 2021) `std::panic!(arbitrary_payload)`.

Alternative names:

- `panic_with!(..)`
- ~~`start_unwind(..)`~~ (panicking doesn't always unwind)
- `throw!(..)`
- `panic_throwing!(..)`
- `panic_with_value(..)`
- `panic_value(..)`
- `panic_with(..)`
- `panic_box(..)`
- `panic(..)`

The equivalent (private, unstable) function in `libstd` is called `std::panicking::begin_panic`.

I suggest `panic_any`, because it allows for any (`Any + Send`) type.

_Tracking issue: #78500_

### CODE REVIEW

**Comment by daboross** [comment]
To bikeshed the name a bit more, what about `panic_with_value`, referring to the fact that one can provide any value? I don't think we necessarily need a short name here, given how infrequently used it is (though if there were a short name which fits well, that would be better).

I'd personally prefer `panic_with_value` over `panic_box` (or `panic_with_box`) just because the fact that it can transmit anything seems slightly more intuitive than the fact that that transmission is done through a box? Though if it did take a `Box<dyn Any + Send + 'static>`, rather than taking `T` and boxing it as an implementation detail, I think the name could fit well.

File: src/libstd/panic.rs [hunk 1]
```diff
@@ -23,6 +23,20 @@ pub use crate::panicking::{set_hook, take_hook};
 #[stable(feature = "panic_hooks", since = "1.10.0")]
 pub use core::panic::{Location, PanicInfo};
 
+/// Panic the current thread, with the given payload as the panic message.
```
Comment (shepmaster) [inline src/libstd/panic.rs:26]:
```suggestion
/// Panic the current thread with the given payload as the panic message.
```

File: src/libstd/panic.rs [hunk 2]
```diff
@@ -23,6 +23,20 @@ pub use crate::panicking::{set_hook, take_hook};
...
+///
+/// This supports an arbitrary panic payload, instead of just (formatted) strings.
```
Comment (shepmaster) [inline src/libstd/panic.rs:28]:
```suggestion
/// This supports an arbitrary panic payload instead of a string.
```

File: src/libstd/panic.rs [hunk 3]
```diff
@@ -23,6 +23,20 @@ pub use crate::panicking::{set_hook, take_hook};
...
+///
+/// The message is attached as a `Box<'static + Any + Send>`, which can be
+/// accessed using [`PanicInfo::payload`].
+///
+/// See the [`panic!`] macro for more information about panicking.
+#[unstable(feature = "panic_box", issue = "none")]
```
Comment (shepmaster) [inline src/libstd/panic.rs:34]: When bikeshedding is done, you'll need to create a tracking issue.

File: src/libstd/panic.rs [hunk 4]
```diff
@@ -23,6 +23,20 @@ pub use crate::panicking::{set_hook, take_hook};
 #[stable(feature = "panic_hooks", since = "1.10.0")]
 pub use core::panic::{Location, PanicInfo};
 
+/// Panic the current thread, with the given payload as the panic message.
+///
+/// This supports an arbitrary panic payload, instead of just (formatted) strings.
+///
+/// The message is attached as a `Box<'static + Any + Send>`, which can be
```
Comment (shepmaster) [inline src/libstd/panic.rs:30]: Attached to what?

**Comment by shepmaster** [comment]
My own 2¢ on the name — I'd want `panic_value!`

**Comment by nikomatsakis** [comment]
Nominating to make @rust-lang/lang members aware of this.

**Comment by nikomatsakis** [comment]
Regarding naming, I agree that something based on "value" is preferable to "box". The fact that boxing occurs doesn't feel like the important thing for users to keep in mind. I'm happy with `panic_with_value`, `panic_value`, or even `panic_with`, although I raised a concern on that last one.

**Comment by nikomatsakis** [comment]
That said, we discussed this in the @rust-lang/lang meeting today and we felt like this sort of bikeshedding is really a @rust-lang/libs affair, so i'm going to retag this as T-libs.

**Comment by joshtriplett** [comment]
As a thought: could we separate the idea of having a separate macro for "panic with this specific value" from the "there's one argument and it's a string literal" case? I'd love to see the 2021-edition `panic!` on a string literal *always* treat it as a format string if it contains any format arguments, and emit a compile-time error if those arguments aren't satisfied. That seems separable from "what do we do if there's one argument and it isn't a string literal".

**Comment by SimonSapin** [comment]
>  to deprecate or disallow (in edition 2021) `std::panic!(arbitrary_payload)`.

That doesn’t seem possible. As far as I know we cannot make macros behave differently based on the edition of the *call site*. (Editions are per-crate, so `std` itself is in one edition.)

**Comment by Mark-Simulacrum** [comment]
It would require compiler magic, I suppose, but only in a more general way, by adding something like `#[deprecated(since = "edition2021")]`. That would be straightforward to add to the compiler I think and is also desirable for the MIN/MAX constants and such.

File: src/libstd/panic.rs (hunk 3 above)
Comment (m-ou-se) [reply src/libstd/panic.rs:34]: Yes. I know.

**Comment by nikomatsakis** [comment]
So, it feels like we need to discuss this PR in the context of the full plan. As I understood it, the plan was this:

* In 2021 Edition:
    * `std::panic!` requires a string literal and literal is treated as a format string
    * `std::panic_with_value!` accepts only a value
* In 2018 Edition:
    * `std::panic!` gets a migration warning if it has no arguments and there is a suggestion to rewrite to `panic_with_value!` -- this is a bit debatable, and we should discuss the particulars, but it seems strictly necessary from my POV to avoid changing runtime semantics

I forget what we would do about `core::panic`...

The hope was indeed that in 2021 Edition crates, the `panic!` macro would always begin with a string literal, and that literal would be treated as a format string. It was mentioned that this could be achieved by having the prelude depend on the edition, such that we had `panic!` route to a distinct macro in the 2021 Edition. It certainly seems possible for us to do it in *some* way, but we'd have to think about the best impl strategy.

**Comment by Mark-Simulacrum** [comment]
`core::panic!` can't accept arbitrary expressions in the single-argument case, only `&str`, so I would suspect something like this:

* In 2021 Edition:
    * `core::panic!` requires a string literal and literal is treated as a format string
         * We may need to explicitly optimize the "identity" case to not bring in format infrastructure, though, for code size reasons.
* In 2018 Edition:
    * `core::panic!("{:?}")` and other single-argument but format strings emit future-compat warnings
               * if you want this, you should write `core::panic!("{}", "{:?}")`  or `core::panic!("{{:?}}")`

**Comment by m-ou-se** [comment]
> We may need to explicitly optimize the "identity" case to not bring in format infrastructure, though, for code size reasons.

That's already taken care of by #74056

**Comment by sfackler** [comment]
Why does `panic_with_value!` need to be a macro at all rather than a normal function?

**Comment by nikomatsakis** [comment]
@sfackler it doesn't, I think that was me just typing wrong

**Comment by Dylan-DPC** [comment]
@m-ou-se if you can resolve the conflicts, we can push this forward

**Comment by m-ou-se** [comment]
This is now part of https://github.com/rust-lang/rfcs/pull/3007

File: library/std/src/panic.rs [hunk 5]
```diff
@@ -23,6 +23,20 @@ pub use crate::panicking::{set_hook, take_hook};
...
 
+/// Panic the current thread with the given message as the panic payload.
+///
+/// The message can be of any (`Any + Send`) type, not just strings.
+///
+/// The message is wrapped in a `Box<'static + Any + Send>`, which can be
+/// accessed later using [`PanicInfo::payload`].
+///
+/// See the [`panic!`] macro for more information about panicking.
+#[unstable(feature = "panic_box", issue = "none")]
```
Comment (camelid) [inline library/std/src/panic.rs:34]: Shouldn't the feature be `panic_any`?

```suggestion
#[unstable(feature = "panic_any", issue = "none")]
```

Comment (m-ou-se) [reply library/std/src/panic.rs:34]: Yes.

**Comment by KodrAus** [comment]
We discussed this in the recent Libs meeting and felt happy with the name `panic_any`. We felt happy to merge this and move discussion over to the RFC itself. With that in mind...

@bors r+ rollup

**Pull request merged**